### PR TITLE
[jsonl] Make skipping entries in the dataloader significantly faster

### PR DESCRIPTION
**Pull request by zdevito**
**Patch Description**
Introduce machinery to skip ahead in the dataset without having to re-tokenize or read the files in the dataset again. 

Measuring in a separate benchmark script in the internal repo indicates this can run reduce the 'fast-forward' stage of data loader from ~20mins to 14 seconds (around 70x speedup).

This works by storing a cache from document idx -> number of tokens that can be stored in a snapshot as a array of numbers. When the token count is known, the `DeferredDataset` creates `DeferredTensor` objects that know their size, how to compute the value if need and how to generate new tensors via slices and concatenates. Another object `SkipDeferredDataset` skips the first `to_skip` elements, without ever computing the DeferredTensors values, bypassing tokenization while keeping the state of the data loader (e.g. the shuffle buffer) exactly the same as if it had actually been running.

### CODE REVIEW

File: metaseq/data/iterators.py [hunk 1]
```diff
@@ -260,29 +265,62 @@ def state_dict(self):
...
+            dataset.len_cache = state_dict["tokenization_cache"]
+            n = state_dict["n"]
+
+            self._itr = self._get_iterator_for_epoch(self.epoch)
+            self._itr.n = n
+            self._itr.sentences_consumed = sentences_consumed
+        else:
+            self._itr = self._get_iterator_for_epoch(self.epoch)
+            # checkpoint from before sentences_consumed was added, slow fast forward...
+            if "iterations_in_epoch" in state_dict and state_dict["iterations_in_epoch"] > 0:
```
Comment (stephenroller) [inline metaseq/data/iterators.py:316]: can we log that we're falling back to slower logic here

Comment (stephenroller) [reply metaseq/data/iterators.py:316]: i like that we can hot swap this in and it will do the right thing on the first try (old logic) and then future restarts get to benefit. that's real nice

File: metaseq/tasks/streaming_language_modeling.py [hunk 2]
```diff
@@ -334,6 +337,7 @@ def load_dataset(self, split: str, epoch=1, combine=False, **kwargs):
 
         dataset = torch.utils.data.ConcatDataset(datasets)
 
+        dataset = DeferredDataset(dataset)
```
Comment (stephenroller) [inline metaseq/tasks/streaming_language_modeling.py:340]: let's leave a few comments explaining the relationship between deferred / skip differed datasets and why we're employing them

**Review by stephenroller — COMMENTED** [review]
v nice

File: metaseq/data/iterators.py [hunk 3]
```diff
@@ -121,24 +121,28 @@ def __init__(self, iterable):
...
             )
         self._peekable_itr = more_itertools.peekable(iterable)
-        self._countable_itr = more_itertools.countable(self._peekable_itr)
+
+        self.num_workers = 1 if num_workers == 0 else num_workers
+        self.batch_size = batch_size
+        self.num_shards = num_shards
+
+        self.n = 0
+        self.sentences_consumed = [0 for _ in range(self.num_workers)]
```
[the remark this answers is not in the document]
Comment (zdevito) [reply metaseq/data/iterators.py:130]: Its the number of 2048-token sequential chunks of tokens (or whatever seq_len is) that each data worker has loaded. Taken from `MAX_SENTENCES` in the benchmark_data_reading script. Ill replace it with whatever the right term is.

File: metaseq/data/iterators.py [hunk 4]
```diff
@@ -121,24 +121,28 @@ def __init__(self, iterable):
...
 
     def __len__(self):
         return 0
 
     def has_next(self):
         return bool(self._peekable_itr)  # whether peekable has items
 
-    @property
-    def n(self):
-        return self._countable_itr.items_seen
```
[the remark this answers is not in the document]
Comment (zdevito) [reply metaseq/data/iterators.py:140]: It is still there, but it is a attribute of the object now. I had to remove the `_countable_itr` because I needed to keep track of the per-worker amount read as well. Doing it as a per-worker count means that we can change the BATCH_SIZE on reload and still restore to the same spot (otherwise, we would have to know all the previous batch sizes to figure out how iteration maps to amount of data we have read through already).

File: metaseq/data/iterators.py [hunk 5]
```diff
@@ -260,29 +265,62 @@ def state_dict(self):
...
+        # will not be copied with the right state
+        if "sentences_consumed" in state_dict and max(state_dict["sentences_consumed"]) > 0:
+            sentences_consumed = state_dict["sentences_consumed"]
+            logger.info(f"Skipping {sentences_consumed} sentences in each worker...")
+            num_workers = 1 if self.num_workers == 0 else self.num_workers
+            assert len(sentences_consumed) == num_workers,\
+                "changing the number of workers in the middle of a shard changes the order the data will be loaded in"
+            dataset = self.dataset
+            while not isinstance(dataset, SkipDeferredDataset):
+                dataset = dataset.dataset
```
[the remark this answers is not in the document]
Comment (zdevito) [reply metaseq/data/iterators.py:303]: Agreed. I think in patch 2, I am going to combine  ShuffleBuffer-GroupDocumentContentIntoBatches-DocumentShuffle into a single dataset object. This (1) improves fast-forward performance a lot by cutting down on byte codes executed, but also (2) means that the skip and defer logic can just be built into this single iterator rather than having to be sandwiched by these extra objects.

**Review by stephenroller — APPROVED** [review]
conditional on s/sentence/sequence/g

File: metaseq/data/deferred.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        t0 = time.time()
+        if isinstance(self.to_skip, int):
+            to_skip = self.to_skip
+        else:
+            info = torch.utils.data.get_worker_info()
+            worker_id = 0 if info is None else info.id
+            to_skip = self.to_skip[worker_id]
+        for i, elem in enumerate(self.dataset):
+            if i >= to_skip:
+                r = tree_map(lambda x: x.realize()
```
[the remark this answers is not in the document]
Comment (ajyl) [reply metaseq/data/deferred.py]: Just curious, why are lambda's not preferred?
